crates/starknet-devnet-core/src/starknet/mod.rs

```diff
@@ -276,10 +280,54 @@ impl Starknet {
         self.predeployed_accounts.get_accounts().to_vec()
     }
 
+    fn update_gas_price(source: Option<NonZeroU128>, target: &mut GasPrice) {
+        if let Some(value) = source {
+            *target = GasPrice(u128::from(value));
+        }
+    }
+
+    fn update_if_none(target: &mut Option<NonZeroU128>, default_value: NonZeroU128) {
+        if target.is_none() {
+            *target = Some(default_value);
+        }
+    }
+
+    fn update_field<T>(source: Option<T>, target: &mut T) {
+        if let Some(value) = source {
+            *target = value;
+        }
+    }
+
     // Update block context
     // Initialize values for new pending block
     pub(crate) fn generate_pending_block(&mut self) -> DevnetResult<()> {
         Self::advance_block_context_block_number(&mut self.block_context);
+
+        match &self.next_block_gas_update {
+            Some(gas_prices) if gas_prices.is_any_field_set() => {
+                Self::update_block_context_gas(&mut self.block_context, gas_prices);
+
+                // Pending block header gas data needs to be updated in case of new prices
+                Self::update_gas_price(
+                    gas_prices.gas_price_wei,
+                    &mut self.blocks.pending_block.header.l1_gas_price.price_in_wei,
+                );
+                Self::update_gas_price(
+                    gas_prices.data_gas_price_wei,
+                    &mut self.blocks.pending_block.header.l1_data_gas_price.price_in_wei,
+                );
+                Self::update_gas_price(
+                    gas_prices.gas_price_strk,
+                    &mut self.blocks.pending_block.header.l1_gas_price.price_in_fri,
+                );
+                Self::update_gas_price(
+                    gas_prices.data_gas_price_strk,
+                    &mut self.blocks.pending_block.header.l1_data_gas_price.price_in_fri,
+                );
```

Comment:
refactored in 1627c07726becc7fdddac2c50973f53eca7dba53

Reply:
I still don't understand why we have to perform the same quadruple assignment in three different places

Reply:
Will check, maybe it's redundant now, at the beginning of this PR gas update data were behaving differently

Reply:
Tried to refactor that to remove `next_block_gas_update` completely but this seems to be problematic because of `BlockContext::new` in `update_block_context_gas()` which is in `generate_pending_block()` problem is that block info needs to be cloned, edited and later reinitialized, this is why this redundancy is needed.

Reply:
Alright, resolving.